src/google/protobuf/compiler/csharp/csharp_message.cc

```diff
@@ -408,9 +405,10 @@ void MessageGenerator::GenerateFrameworkMethods(io::Printer* printer) {
         "}\n\n");
 
     // GetHashCode
+    // Start with a non-zero value to easily distinguish between null and "empty" messages.
```

Comment:
I'd be happy to use 1 if that's more common in your experience. I think it makes sense to use a different "empty" hash code for "message", "repeated field" and "map field" though - are 1, 2, 3 fine for you?

(Note that this PR depends on the map PR, so it may be worth reviewing that first.)